### ISSUE TITLE
Migrate the main parser to Menhir

### ISSUE THREAD

**Issue by dhil** [links-lang/links#411]
Our main parser / grammar `parser.mly` could do with some cleaning.

Some of the production rules such as `preamble` are essentially garbage. We should get rid of rules like those. In general we should probably simplify/refactor our grammar. The best solution might be to migrate the main parser from ocamlyacc to Menhir, since Menhir provides some convenient combinators that enable us to get rid of most (if not all) of the `_list` and `_perhaps` productions.

In doing so, we might want to consider whether my favourite Links program should remain syntactically valid
```links
links> {;};
() : ()
```

**Comment by frank-emrich** [comment]
I just saw that Menhir has built-in support for grammar modularity (cf. [http://gallium.inria.fr/~fpottier/menhir/manual.html#sec%3Asplit]). This means that the definition of the overall  grammar can be split over multiple files and rules for shared non-terminals are merged. Other parser generators have more advanced modularity features, but Menhir's basic support for modularity is certainly nice to have!

**Comment by jamescheney** [comment]
I don't have the feeling that this is high on anyone's to-do list at the moment.  It also seems like #411 ought to be mostly orthogonal to #187, so as long as people aren't working on them at the same time there shouldn't be a problem.  (Also, #187 seems higher priority and might give us useful information about how / if we want to move to Menhir / refactor the parser)

**Comment by dhil** [comment]
I agree with @jamescheney. Migration to Menhir is unlikely to happen anytime soon. Though, I firmly believe that Menhir would simplify and generally make it easier to work with our parser(s).

**Comment by jstolarek** [comment]
One nice feature to have would be the ability to record source positions of various elements of a production. Say I have a production:

```
foo:
| A B C       { ... }
```

Right now we only have access to `pos()`, which I understand records positon for the whole production. I would like to be able to say, e.g. `pos $2` to record position of `B`.

A separate questions are: 1) which elements of a parse tree should be annotated with position and; 2) how productions should attach position information. Re 1 I think it would make sense to be consistent and produce a fully annotated tree, i.e. every parsed element should carry a location information. Re 2, parser would probably be easier to understand and manage if we had a convention that every production returns an annotated element. So instead of having:

```
foo:
| A   { Foo }
| B   { Bar }

bar:
| C   { Quux }
| D   { Quuz }

baz:
| foo bar   { Baz ($1, pos $1) ($2, pos $2) }
```

we should have:

```
foo:
| A   { Foo, pos() }
| B   { Bar, pos() }

bar:
| C   { Quux, pos() }
| D   { Quuz, pos() }

baz:
| foo bar   { Baz $1 $2, pos () }
```

This is more verbose, but it should be easier to keep track of how productions work and what they return. Incidentally, this approach also eliminates a need to record positions of individual elements of a production I mentioned above.

The two questions above seem independent of whether we decide to use Menhir or not, but perhaps Menhir has some features that would make position tracking easier?

**Comment by slindley** [comment]
Migrating to Menhir seems like an uncontroversial change that will almost certainly be a benefit. While it isn't high on our current priority list, if someone wanted to give it a go then that would certainly be worthwhile.

**Comment by slindley** [comment]
I agree that making the parser more uniform in the manner that you describe is a good idea.

**Comment by jstolarek** [comment]
I started to work on this yesterday to see how difficult this is and the migration turns out to be fairly straightforward, except maybe for a ton of boilerplate changes to correctly handle source positions.  However, there is one thing that I need to consult.
 
Menhir is more strict about handling end-of-stream conflicts and reports warnings about two such conflicts in our grammar. I was able to silence one of them by just refactoring productions, but dealing with another actually requires modifying Links grammar. Here's what happens. Say we have a Links program:

```
open Foo
```

This is parsed using productions `interactive -> nofun_declaration -> links_open`. `links_open` after the changes I've made looks like this:

```
links_open:
| OPEN separated_nonempty_list(DOT, CONSTRUCTOR) { with_pos (`QualifiedImport  $2 ) (pos()) }
```

There are two ways this production can behave when parsing `open Foo`:

1. if this is the end of file then reduce, production ends;
2. if `Foo` is followed by a dot then shift, keep parsing.

To know how to act a parser needs to look at the next character, but if we are reading from a stream there is a risk that no further input is available (first case above) the program will block waiting on input - hence Menhir's warning. A solution is to introduce a symbol that marks the end production. My idea is to require a semicolon after an `open` declaration:

```
open Foo;
```

This resolves the conflict and doesn't seem too bad - almost all `nofun_declarations` end with a semicolon anyway. I wonder how others feel about this change.

**Comment by slindley** [comment]
A semicolon after an `open` declaration sounds fine to me - and more in keeping with the rest of Links syntax anyway.
